lib/evmone/instructions.hpp

```diff
@@ -91,75 +108,87 @@ inline bool check_memory(ExecutionState& state, const uint256& offset, const uin
+/// - the "base" gas const has already been charged,
+/// - the `stack` pointer points to the EVM stack top element.
+/// Moreover, these implementations _do not_ inform about new stack height
+/// after execution. The adjustment must be performed by the caller.
+namespace core
 {
-    return {EVMC_SUCCESS};
+inline void noop(StackTop /*stack*/) noexcept {}
+inline constexpr auto pop = noop;
+inline constexpr auto jumpdest = noop;
+
+template <evmc_status_code Status>
+inline StopToken stop_impl() noexcept
+{
+    return {Status};
 }
+inline constexpr auto stop = stop_impl<EVMC_SUCCESS>;
+inline constexpr auto invalid = stop_impl<EVMC_INVALID_INSTRUCTION>;
 
-inline void add(ExecutionState& state) noexcept
+inline void add(StackTop stack) noexcept
 {
-    state.stack.top() += state.stack.pop();
+    stack.top() += stack.pop();
 }
 
-inline void mul(ExecutionState& state) noexcept
+inline void mul(StackTop stack) noexcept
 {
-    state.stack.top() *= state.stack.pop();
+    stack.top() *= stack.pop();
 }
 
-inline void sub(ExecutionState& state) noexcept
+inline void sub(StackTop stack) noexcept
 {
-    state.stack[1] = state.stack[0] - state.stack[1];
-    state.stack.pop();
+    stack[1] = stack[0] - stack[1];
 }
 
-inline void div(ExecutionState& state) noexcept
+inline void div(StackTop stack) noexcept
 {
-    auto& v = state.stack[1];
-    v = v != 0 ? state.stack[0] / v : 0;
-    state.stack.pop();
+    auto& v = stack[1];
+    v = v != 0 ? stack[0] / v : 0;
 }
 
-inline void sdiv(ExecutionState& state) noexcept
+inline void sdiv(StackTop stack) noexcept
 {
-    auto& v = state.stack[1];
-    v = v != 0 ? intx::sdivrem(state.stack[0], v).quot : 0;
-    state.stack.pop();
+    auto& v = stack[1];
+    v = v != 0 ? intx::sdivrem(stack[0], v).quot : 0;
 }
 
-inline void mod(ExecutionState& state) noexcept
+inline void mod(StackTop stack) noexcept
 {
-    auto& v = state.stack[1];
-    v = v != 0 ? state.stack[0] % v : 0;
-    state.stack.pop();
+    auto& v = stack[1];
+    v = v != 0 ? stack[0] % v : 0;
 }
 
-inline void smod(ExecutionState& state) noexcept
+inline void smod(StackTop stack) noexcept
 {
-    auto& v = state.stack[1];
-    v = v != 0 ? intx::sdivrem(state.stack[0], v).rem : 0;
-    state.stack.pop();
+    auto& v = stack[1];
+    v = v != 0 ? intx::sdivrem(stack[0], v).rem : 0;
 }
 
-inline void addmod(ExecutionState& state) noexcept
+inline void addmod(StackTop stack) noexcept
 {
-    const auto x = state.stack.pop();
-    const auto y = state.stack.pop();
-    auto& m = state.stack.top();
+    const auto& x = stack.pop();
+    const auto& y = stack.pop();
+    auto& m = stack.top();
     m = m != 0 ? intx::addmod(x, y, m) : 0;
 }
 
-inline void mulmod(ExecutionState& state) noexcept
+inline void mulmod(StackTop stack) noexcept
 {
-    const auto x = state.stack.pop();
-    const auto y = state.stack.pop();
-    auto& m = state.stack.top();
+    const auto& x = stack[0];
```

Comment:
I changed only this implementation because this gives slightly better assembly. All other changes make no difference except to symbol names.